# Dimensionality reduction and PCA

## 1. Discuss curse of dimensionality

--As the number of features increases the efficiency of the algorithm is generally decreases as the number of dimensions increases it requires high computation requirements.
--It is also possible that all features are are not equally important for  predictiction in our model.

## 2. Discuss any 3 dimensionality reduction techniques

- Principal component analysis (PCA)
- Linear Discriminatory Analysis (LDA)
- Singular Value Composition

## 3. Explain PCA in your own words

PCA is used in exploratory data analysis and for making predictive models. It is commonly used for dimensionality reduction by 
projecting each data point onto only the first few principal components to obtain lower-dimensional data while preserving as much
of the data's variation as possible.

## 4. Perform classification on wine dataset using any model of your choice using all the features

In [56]:
wine_dataset = datasets.load_wine()
X=wine_dataset.data
y=wine_dataset.target

In [58]:
import xgboost as xgb
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix 
from sklearn.metrics import classification_report

X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.35,random_state=36)
xgb_reg = xgb.XGBClassifier()
xgb_reg.fit(X_train, y_train)

y_test_predict = xgb_reg.predict(X_test)

print(confusion_matrix(y_test, y_test_predict))
print(classification_report(y_test, y_test_predict))

[11:59:54] WARNING: C:/Users/Administrator/workspace/xgboost-win64_release_1.5.1/src/learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'multi:softprob' was changed from 'merror' to 'mlogloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[[23  0  0]
 [ 1 22  1]
 [ 0  0 16]]
              precision    recall  f1-score   support

           0       0.96      1.00      0.98        23
           1       1.00      0.92      0.96        24
           2       0.94      1.00      0.97        16

    accuracy                           0.97        63
   macro avg       0.97      0.97      0.97        63
weighted avg       0.97      0.97      0.97        63



C:\Users\Checkout\anaconda3\lib\site-packages\xgboost\sklearn.py:1224: UserWarning: The use of label encoder in XGBClassifier is deprecated and will be removed in a future release. To remove this warning, do the following: 1) Pass option use_label_encoder=False when constructing XGBClassifier object; and 2) Encode your labels (y) as integers starting with 0, i.e. 0, 1, 2, ..., [num_class - 1].
  warnings.warn(label_encoder_deprecation_msg, UserWarning)


## 5. Perform classification on wine dataset using PCA and taking all top components such that 90% variance is explained by them

In [ ]:
# importing the required libraries
import pandas as pd 
import numpy as np
import matplotlib.pyplot as plt
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler
from sklearn import preprocessing
from sklearn import datasets

In [ ]:
#url = 'http://archive.ics.uci.edu/ml/machine-learning-databases/wine-quality/winequality-red.csv'
#df= pd.read_csv(wine_dataset, sep=";")
#df= pd.read_csv(wine_dataset)
#df.head()

In [55]:
#X=wine_dataset.data
#y=wine_dataset.target

In [ ]:
# scaling the dataset
X = StandardScaler().fit_transform(X)

In [60]:
from sklearn.decomposition import PCA

pca = PCA(n_components = 8) # creates an instance of PCA class
principalComponents = pca.fit_transform(X)
principalDf = pd.DataFrame(data = principalComponents, columns = ['principal component 1', 'principal component 2','principal component 3','principal component 4','principal component 5','principal component 6','principal component 7','principal component 8'])

In [ ]:
# creating the final dataframe with 2 principal components and the target variable
#finalDf = pd.concat([principalDf, df[['quality']]], axis = 1)

In [62]:
# using xgb boost as it is very fast and highly efficient.
import xgboost as xgb
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix 
from sklearn.metrics import classification_report

X_train, X_test, y_train, y_test = train_test_split(principalDf,y,test_size=0.25,random_state=48)
xgb_reg = xgb.XGBClassifier()
xgb_reg.fit(X_train, y_train)

y_test_predict = xgb_reg.predict(X_test)

print(confusion_matrix(y_test, y_test_predict))
print(classification_report(y_test, y_test_predict))

[12:04:42] WARNING: C:/Users/Administrator/workspace/xgboost-win64_release_1.5.1/src/learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'multi:softprob' was changed from 'merror' to 'mlogloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[[ 9  0  0]
 [ 2 17  0]
 [ 0  2 15]]
              precision    recall  f1-score   support

           0       0.82      1.00      0.90         9
           1       0.89      0.89      0.89        19
           2       1.00      0.88      0.94        17

    accuracy                           0.91        45
   macro avg       0.90      0.93      0.91        45
weighted avg       0.92      0.91      0.91        45



C:\Users\Checkout\anaconda3\lib\site-packages\xgboost\sklearn.py:1224: UserWarning: The use of label encoder in XGBClassifier is deprecated and will be removed in a future release. To remove this warning, do the following: 1) Pass option use_label_encoder=False when constructing XGBClassifier object; and 2) Encode your labels (y) as integers starting with 0, i.e. 0, 1, 2, ..., [num_class - 1].
  warnings.warn(label_encoder_deprecation_msg, UserWarning)


## 6. Discuss the difference in results vs difference in number of features used for classification.

From reducing features from 13(accuracy =0.97) to 8( accuracy =0.91) we reduce the computational reqirements and we even get the similar accuracy only.We also avoided our model from getting overfitted.